### PR TITLE
Fix overlay recalculation bug with multiple batches

### PR DESCRIPTION
**Pull request by wu-hui**
Fixes: https://github.com/firebase/firebase-android-sdk/issues/3490 and https://github.com/FirebaseExtended/flutterfire/issues/7988

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java [hunk 1]
```diff
@@ -165,6 +165,11 @@ private void recalculateAndSaveOverlays(Map<DocumentKey, MutableDocument> docs)
     // along the way.
     for (MutationBatch batch : batches) {
       for (DocumentKey key : batch.getKeys()) {
+        // If this batch has documents not included in passed in `docs`, skip them.
+        if (!docs.containsKey(key)) {
+          continue;
+        }
+
         FieldMask mask = masks.containsKey(key) ? masks.get(key) : FieldMask.EMPTY;
         mask = batch.applyToLocalView(docs.get(key), mask);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java:174]: Can we change this to only issue a single lookup in `docs` and use the nullable return value as the exit criteria?

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java [hunk 2]
```diff
@@ -306,4 +306,31 @@ public void testRunBatch() {
     assertTrue(snapshot.exists());
     assertEquals(map("foo", "bar", "baz", 42L), snapshot.getData());
   }
+
+  @Test
+  public void testCanHandlePendingMutationsWithIrrelevantDocuments() {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java:311]: Can you use a different term for `irrelevant` here? 

`testCanHandleBatchThatCausesPartialViewRecalculation()` (maybe)

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java [hunk 3]
```diff
@@ -306,4 +306,31 @@ public void testRunBatch() {
...
+    DocumentReference doc = testDocument();
+    DocumentReference anotherDoc = doc.getParent().document();
+    FirebaseFirestore db = doc.getFirestore();
+    // Prepare two batches, the first one will get rejected by the backend.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java:315]: Empty line before?

**Review by schmidt-sebastian — COMMENTED** [review]
Please add changelog

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java (hunk 1 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java]: Done.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java (hunk 2 above)
Comment (wu-hui) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java]: Renamed.

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java [hunk 4]
```diff
@@ -306,4 +306,31 @@ public void testRunBatch() {
...
+    batch2.commit();
+
+    waitFor(doc.getFirestore().waitForPendingWrites());
+    DocumentSnapshot snapshot = waitFor(doc.get());
+    assertTrue(snapshot.exists());
+    assertEquals(map("foo", "bar_set"), snapshot.getData());
+
+    snapshot = waitFor(anotherDoc.get());
+    assertTrue(snapshot.exists());
+    assertEquals(map("foo", "another"), snapshot.getData());
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/androidTest/java/com/google/firebase/firestore/WriteBatchTest.java]: Done.

File: firebase-firestore/CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,8 @@ by opting into a release at
   the shutdown from proceeding if a network connection was opened right before.
 - [changed] Queries are now send to the backend before the SDK starts local 
   processing, which reduces overall Query latency.
+- [fixed] Fixed an issue where mutations with multiple documents are not handled
+  correctly during previous mutation acknowledgement(#3490).
```
Comment (schmidt-sebastian) [inline firebase-firestore/CHANGELOG.md:14]: Maybe mention that this threw a NPE (sometimes, not always)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java [hunk 6]
```diff
@@ -165,8 +165,14 @@ private void recalculateAndSaveOverlays(Map<DocumentKey, MutableDocument> docs)
     // along the way.
     for (MutationBatch batch : batches) {
       for (DocumentKey key : batch.getKeys()) {
+        // If this batch has documents not included in passed in `docs`, skip them.
+        MutableDocument baseDoc = docs.get(key);
+        if (baseDoc == null) {
+          continue;
+        }
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java:172]:
```suggestion
        MutableDocument baseDoc = docs.get(key);
        if (baseDoc == null) { 
          // If this batch has documents not included in passed in `docs`, skip them.
          continue;
        }
```

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java [hunk 7]
```diff
@@ -1599,4 +1599,21 @@ public void testOnlyPersistsUpdatesForDocumentsWhenVersionChanges() {
...
+    // See: https://github.com/firebase/firebase-android-sdk/issues/3490
+    writeMutation(patchMutation("foo/bar", map("foo", "bar")));
+    writeMutations(
+        asList(
+            setMutation("foo/bar", map("foo", "bar-set")),
+            setMutation("foo/another", map("foo", "another"))));
+
+    rejectMutation();
+    assertContains(doc("foo/bar", 0, map("foo", "bar-set")).setHasLocalMutations());
+    assertContains(doc("foo/another", 0, map("foo", "another")).setHasLocalMutations());
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/LocalStoreTestCase.java:1617]: Nice!

**Review by schmidt-sebastian — APPROVED** [review]

File: firebase-firestore/CHANGELOG.md (hunk 5 above)
Comment (wu-hui) [reply firebase-firestore/CHANGELOG.md]: Done.

**Pull request merged**